### PR TITLE
Improved BaseReporter (replaces #896)

### PR DESCRIPTION
**Pull request by georgecrawford**
Reworking of https://github.com/webdriverio/webdriverio/pull/896, including:

- Add command hooks to enable reporting of `beforeCommand`, `afterCommand` and `screenshot` (cd8887f)
- Rework base reporter to be more useful, and allow for concurrent specs (0be45bf)
- Report screenshot file paths (0be45bf)
- Report `app` when sanitizing capabilities (2a85761)
- Enable `runner:command` and `runner:result` to be associated with a test (7057bdf)
- Expose upstream errors when failing to load wdio framework or plugin (2c02405)

### CODE REVIEW

File: lib/utils/BaseReporter.js [hunk 1]
```diff
@@ -173,6 +386,29 @@ class BaseReporter extends events.EventEmitter {
         return `\u001b[${COLORS[type]}m${str}\u001b[0m`
     }
 
+    limit (val) {
+        switch (Object.prototype.toString.call(val)) {
+        case '[object String]':
+            // Check the first 100 characters to detect a base64 string: http://stackoverflow.com/a/475217/67190
```
Comment (christian-bromann) [inline lib/utils/BaseReporter.js]: I just merged #990. @sbmaxx used the [validator](https://www.npmjs.com/package/validator) package to detect base64, let's do that here too.

Comment (georgecrawford) [inline lib/utils/BaseReporter.js]: Nice, thanks.

Comment (georgecrawford) [inline lib/utils/BaseReporter.js]: Done

**Comment by georgecrawford** [comment]
@christian-bromann can you copy over the line comments from https://github.com/webdriverio/webdriverio/pull/896 which still apply please? I'm a little confused about what you'd like to be changed.

**Comment by christian-bromann** [comment]
There is nothing to change. Looks awesome! Please merge it. Thanks

**Comment by georgecrawford** [comment]
Woohoo!!!

**Pull request merged**